source/utilities.tex

```diff
@@ -19111,3 +19112,86 @@
 Otherwise, \tcode{endian::native} is not equal
 to either \tcode{endian::big} or \tcode{endian::little}.
 \end{itemdescr}
+
+\rSec1[debugging]{Debugging}
+
+\rSec2[debugging.general]{General}
+
+\pnum
+This subclause \ref{debugging} describes functionality to introspect and
+interact with the execution of the program.
+
+\begin{note}
+The facilities provided by the debugging functionality interact with a program
+that may be tracing the execution of a C++ program, such as a debugger.
+\end{note}
+
+\rSec2[debugging.syn]{Header \tcode{<debugging>} synopsis}
+
+\begin{codeblock}
+// all freestanding
+namespace std {
+	// \ref{debugging.utility}, utility
+	void breakpoint() noexcept;
+	void breakpoint_if_debugging() noexcept;
+	bool is_debugger_present() noexcept;
+}
+\end{codeblock}
+
+\rSec2[debugging.utility]{Utility}
+
+\indexlibraryglobal{breakpoint}%
+\begin{itemdecl}
+void breakpoint() noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+
+\pnum
+The semantics of this function are implementation-defined.
+
+\begin{note}
+When invoked the execution of the program temporarily halts and execution is
+handed to the debugger until such a time as: the program is terminated by the
+debugger or, the debugger resumes execution of the program as if the function
+was not invoked.
+\end{note}
+
+\end{itemdescr}
+
+\indexlibraryglobal{breakpoint_if_debugging}%
+\begin{itemdecl}
+void breakpoint_if_debugging() noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+
+\pnum
+\effects
+Equivalent to:
+\begin{codeblock}
+if (is_debugger_present()) breakpoint();
+\end{codeblock}
+
+\end{itemdescr}
+
+\indexlibraryglobal{is_debugger_present}%
+\begin{itemdecl}
+bool is_debugger_present() noexcept;
+\end{itemdecl}
+
+\begin{itemdescr}
+
+\pnum
+The semantics of this function are implementation-defined.
+
+\begin{note}
+When tracing with a debugger the execution of a program an implementation
+returns \tcode{true}. An implementation performs an immediate query, as needed,
+to determine if the program is traced by a debugger. On Windows, or equivalent,
+systems this can be achieved by calling the \tcode{::IsDebuggerPresent()} Win32
+function. On POSIX this can be achieved by checking for a tracer parent process,
+with best effort determination that such a tracer parent process is a debugger. 
```

Comment:
```suggestion
When tracing the execution of a program with a debugger, an implementation
returns \tcode{true}. An implementation performs an immediate query, as needed,
to determine if the program is traced by a debugger. On Windows or equivalent
systems, this can be achieved by calling the \tcode{::IsDebuggerPresent()} Win32
function. On POSIX, this can be achieved by checking for a tracer parent process,
with best effort determination that such a tracer parent process is a debugger. 
```